devtools/gamma-server/test/gamma_server_and_cli_test.go

```diff
@@ -0,0 +1,276 @@
+		Arguments:    []gammacli.JSONMethodArgument{},
+	}
+
+	callJSONBytes, _ := json.Marshal(&getCounterJSON)
+	return callJSONBytes
+}
+
+func generateAddCounterJSON(amount uint64) []byte {
+	arg := gammacli.JSONMethodArgument{Name: "amount", Type: "uint64", Value: amount}
+
+	addAmountToCounterJSON := &gammacli.JSONTransaction{
+		ContractName: "Counter",
+		MethodName:   "add",
+		Arguments:    []gammacli.JSONMethodArgument{arg},
+	}
+
+	addJSONBytes, _ := json.Marshal(&addAmountToCounterJSON)
+	return addJSONBytes
+}
+
+func TestGammaFlowWithActualJSONFilesUsingBenchmarkToken(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping e2e tests in short mode")
+	}
+
+	port := ":8080"
+	gamma := gammacli.StartGammaServer(port, false)
+	defer gamma.GracefulShutdown(1 * time.Second)
+
+	time.Sleep(100 * time.Millisecond) // wait for server to start
+
+	keyPair := keys.Ed25519KeyPairForTests(0)
+	targetAddress := builders.AddressForEd25519SignerForTests(2)
+
+	transferJSONBytes := generateTransferJSON(42, targetAddress)
+
+	err := ioutil.WriteFile("../json/transfer.json", transferJSONBytes, 0644)
+	if err != nil {
+		fmt.Println("Couldn't write file", err)
+	}
+	require.NoError(t, err, "Couldn't write transfer JSON file")
+
+	baseCommand := ClientBinary()
+	sendCommand := append(baseCommand,
+		"run", "send", "../json/transfer.json",
+		"-public-key", keyPair.PublicKey().String(),
+		"-private-key", keyPair.PrivateKey().String())
+
+	sendCommandOutput := runCommand(sendCommand, t)
+
+	response := &sendTransactionCliResponse{}
+	unmarshalErr := json.Unmarshal([]byte(sendCommandOutput), &response)
+
+	require.NoError(t, unmarshalErr, "error unmarshal cli response")
+	require.Equal(t, 1, response.TransactionReceipt.ExecutionResult, "JSONTransaction status to be successful = 1")
+	require.Equal(t, 1, response.TransactionStatus, "JSONTransaction status to be successful = 1")
+	require.NotNil(t, response.TransactionReceipt.Txhash, "got empty txhash")
+
+	getBalanceJSONBytes := generateGetBalanceJSON(targetAddress)
+	err = ioutil.WriteFile("../json/getBalance.json", getBalanceJSONBytes, 0644)
+	if err != nil {
+		fmt.Println("Couldn't write file", err)
+	}
+	require.NoError(t, err, "Couldn't write getBalance JSON file")
+
+	getCommand := append(baseCommand, "run", "call", "../json/getBalance.json")
+
+	callOutputAsString := runCommand(getCommand, t)
+	fmt.Println(callOutputAsString)
+
+	callResponse := &callMethodCliResponse{}
+	callUnmarshalErr := json.Unmarshal([]byte(callOutputAsString), &callResponse)
+
+	require.NoError(t, callUnmarshalErr, "error calling call_method")
+	require.Equal(t, 0, callResponse.CallResult, "Wrong callResult value")
+	require.Len(t, callResponse.OutputArguments, 1, "expected exactly one output argument returned from getBalance")
+	require.EqualValues(t, uint64(42), uint64(callResponse.OutputArguments[0].Value.(float64)), "expected balance to equal 42")
+}
+
+func TestGammaCliDeployWithUserDefinedContract(t *testing.T) {
```

Comment:
this is a long test that tells a story - consider extracting each step in the story to a separate function with semantic meaning.

Reply:
Done